### PR TITLE
feat: add hook data 

### LINKED ISSUE #1472: Add Hook Data
## Summary

The OpenFeature specification now includes support for hook data, which allows hooks to maintain state across their execution stages (before → after/error → finally). This feature enables use cases like distributed tracing, performance monitoring, and multi-stage validation.

This issue tracks the implementation of hook data support in this SDK to achieve compliance with the latest specification.

## Action Required

SDK maintainers need to implement hook data support according to the specification defined in:
- https://github.com/open-feature/spec/pull/273 - Initial hook data specification
- https://github.com/open-feature/spec/pull/307 - Hook data updates

A reference implementation is available in the .NET SDK PR: https://github.com/open-feature/dotnet-sdk/pull/387

## Hook data specification enables stateful hook operations

The OpenFeature hook data specification provides a mechanism for different stages of the same hook to share state during a single flag evaluation. This feature supports implementing telemetry, performance monitoring, and resource management use cases. The .NET SDK's implementation in https://github.com/open-feature/dotnet-sdk/pull/387 demonstrates a successful approach that other SDKs can follow.

## Reference Links

- **Specification PRs**: 
  - https://github.com/open-feature/spec/pull/273 - Initial hook data specification
  - https://github.com/open-feature/spec/pull/307 - Hook data updates
- **Implementation Example**: https://github.com/open-feature/dotnet-sdk/pull/387
- **Documentation**:
  - [OpenFeature Hook Specification](https://openfeature.dev/specification/sections/hooks/)
  - [Hook Concepts Documentation](https://openfeature.dev/docs/reference/concepts/hooks/)

## Understanding hook data fundamentals

Hook data provides the ability for hooks to maintain state across their execution lifecycle. When a hook starts an operation in the `before` stage (like opening a telemetry span), it needs a way to access that same data in the `after` stage to complete the operation. Without hook data, developers resort to workarounds like global state, weak maps, or thread-local storage, which introduce complexity and potential bugs.

The specification defines hook data as a **mutable, per-hook, per-evaluation** data structure. This scoping means each hook instance gets its own isolated data store that persists only for the duration of a single flag evaluation. Different hooks evaluating the same flag don't share data, preventing interference and ensuring predictable behavior.

### Data structure and access patterns

Hook data **MUST** be implemented as a structure supporting arbitrary properties with string keys and values of any type. The structure must be mutable to allow stages to modify shared state. SDKs typically implement this as a map-like interface with `set(key, value)` and `get(key)` methods.

The data structure must be created before the first hook stage executes and remain available throughout all stages (before → after/error → finally). This lifecycle management is the SDK's responsibility, not the hook developer's.

### Isolation and scoping guarantees

Each hook instance maintains its own hook data - there's no sharing between different hooks. This isolation extends to:
- Different hook types running for the same evaluation
- Multiple instances of the same hook type
- Hooks registered at different levels (global, client, invocation)

The specification explicitly states that hook data is not shared between different hooks, ensuring separation of concerns.

### Integration with hook context

Hook data must be accessible through the hook context parameter passed to all hook stages. The typical pattern follows:

```pseudocode
context.hookData.set("key", value)  // In before stage
value = context.hookData.get("key") // In after stage
```

## Learning from the .NET SDK implementation

The .NET SDK's PR https://github.com/open-feature/dotnet-sdk/pull/387 provides implementation insights. Their approach enhanced the existing `HookContext<T>` class with a `Data` property, maintaining backward compatibility while adding new functionality. Lessons include:

**API Design**: The implementation uses a simple key-value store pattern that's intuitive for developers. Type safety is handled through casting when retrieving values, following .NET conventions.

**No Breaking Changes**: Existing hooks continue working without modification. The hook data feature is additive - hooks can opt-in to using it without requiring changes to hooks that don't need state sharing.

**Examples**: The implementation includes examples like timing hooks that measure execution duration, demonstrating the feature's value to developers.

### Phase 1: Core implementation

Start by implementing the basic hook data structure with these capabilities:
- String-keyed storage with any-type values
- Mutable data access through set/get operations
- Proper lifecycle management (creation before first stage, cleanup after finally)
- Thread-safe access for concurrent evaluations

### Phase 2: Integration and testing

Integrate hook data with your existing hook infrastructure:
- Add data property to hook context
- Ensure data flows through all hook stages
- Implement isolation between hook instances
- Create comprehensive test suites covering edge cases

### Phase 3: Documentation and examples

Provide guidance for hook developers:
- API documentation with type signatures
- Examples (timing, telemetry, validation)
- Best practices and anti-patterns
- Migration guide if introducing any changes

### Memory management

Hook data should be garbage collected after the evaluation completes. Consider implementing weak references or explicit cleanup to prevent memory leaks in long-running applications.

### Thread safety

Many applications evaluate flags concurrently. Ensure hook data access is thread-safe without introducing significant performance overhead. Consider using concurrent data structures appropriate for your language.

### Type safety

In statically-typed languages, consider providing type-safe access patterns while maintaining the flexibility of arbitrary value storage. Generic methods or type assertions can help balance safety with usability.

### Performance impact

Hook data should add minimal overhead to flag evaluations. Consider lazy initialization - only create the data structure when a hook actually uses it. Benchmark the implementation to ensure acceptable performance.

### OpenTelemetry integration

The canonical use case involves creating spans in the `before` stage and ending them in `after`:

```java
// Before stage
Span span = tracer.spanBuilder("feature-flag-evaluation").startSpan();
context.hookData.set("span", span);

// After stage
Span span = (Span) context.hookData.get("span");
span.end();
```

### Performance monitoring

Tracking evaluation duration is a common requirement:

```csharp
// Before stage
context.Data.Set("startTime", DateTime.UtcNow);

// After stage
var duration = DateTime.UtcNow - (DateTime)context.Data.Get("startTime");
logger.LogInformation($"Evaluation took {duration.TotalMilliseconds}ms");
```

### Multi-stage validation

Complex validation scenarios benefit from accumulating results:

```python
# Before stage
context.hook_data.set("validation_errors", [])

# Multiple validation stages can add errors
errors = context.hook_data.get("validation_errors")
if not is_valid(context):
    errors.append("Validation failed")

# Finally stage checks accumulated errors
if context.hook_data.get("validation_errors"):
    logger.warning(f"Validation issues: {errors}")
```

## Testing requirements

Comprehensive test coverage should include:

1. **Basic functionality**: Set and get operations work correctly
2. **Isolation**: Data doesn't leak between hook instances
3. **Lifecycle**: Data persists through all stages but not beyond
4. **Type handling**: Various data types can be stored and retrieved
5. **Edge cases**: Null values, missing keys, concurrent access
6. **Performance**: Minimal overhead for evaluations not using hook data

## Migration and compatibility

When implementing hook data support:

1. **Maintain backward compatibility**: Existing hooks must continue working
2. **Version appropriately**: Follow semantic versioning based on impact
3. **Provide migration guides**: Help developers adopt the new feature
4. **Consider feature flags**: Allow gradual rollout of hook data support

## Success criteria

A complete hook data implementation should:
- Pass all specification compliance tests
- Support the primary use cases (telemetry, monitoring, validation)
- Add minimal performance overhead
- Provide good developer experience
- Include comprehensive documentation
- Maintain backward compatibility

## Conclusion

Hook data provides a feature that enables observability and monitoring capabilities in OpenFeature. By following the specification requirements and learning from successful implementations like the .NET SDK, maintainers can provide consistent, reliable hook data support across all OpenFeature SDKs. The implementation should balance flexibility with safety, performance with functionality, and maintain the isolation guarantees that make hook data predictable and useful.

## Additional Resources

- https://github.com/open-feature/spec
- [Hook Specification Details](https://openfeature.dev/specification/sections/hooks/)
- [Hook Concepts Guide](https://openfeature.dev/docs/reference/concepts/hooks/)
- [.NET SDK Documentation](https://openfeature.dev/docs/reference/technologies/server/dotnet/)
- https://github.com/open-feature/dotnet-sdk/releases

### PR DESCRIPTION
**Pull request by alexandraoberaigner**
## This PR
improves PR #1506 from @mdxabu  who started the work on adding hook data

### Related Issues

Fixes #1472

### Notes
Tests & Impl are still WIP: 
* HookData isolation might not fully work yet (not sure if the test is setup incorrectly or we need to adapt the implementation still)
* pls let me know if you see any edge cases that I didn't think of yet 🙏

### CODE REVIEW

File: src/main/java/dev/openfeature/sdk/HookSupport.java [hunk 1]
```diff
@@ -5,62 +5,92 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
-import java.util.function.Consumer;
+import java.util.function.BiConsumer;
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
```
Comment (chrfwow) [inline src/main/java/dev/openfeature/sdk/HookSupport.java]: This makes the build fail: 
```
Error:  Used undeclared dependencies found:
Error:     org.apache.commons:commons-lang3:jar:3.18.0:provided
```
Can we use something else instead?

File: src/main/java/dev/openfeature/sdk/HookSupport.java [hunk 2]
```diff
@@ -71,26 +101,51 @@ private <T> void executeChecked(Hook<T> hook, Consumer<Hook<T>> hookCode, String
...
+        // Create hook data for each hook instance
+        Map<Hook, HookData> hookDataMap = new HashMap<>();
         for (Hook hook : reversedHooks) {
             if (hook.supportsFlagValueType(flagValueType)) {
-                Optional<EvaluationContext> optional =
-                        Optional.ofNullable(hook.before(hookCtx, hints)).orElse(Optional.empty());
+                hookDataMap.put(hook, HookData.create());
+            }
+        }
+        */
```
Comment (chrfwow) [inline src/main/java/dev/openfeature/sdk/HookSupport.java]: Can be removed

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java [hunk 3]
```diff
@@ -164,29 +165,21 @@ private <T> FlagEvaluationDetails<T> evaluateFlag(
...
-            var stateManager = openfeatureApi.getFeatureProviderStateManager(this.domain);
+            final var stateManager = openfeatureApi.getFeatureProviderStateManager(this.domain);
             // provider must be accessed once to maintain a consistent reference
-            var provider = stateManager.getProvider();
-            var state = stateManager.getState();
-
+            final var provider = stateManager.getProvider();
+            final var state = stateManager.getState();
+            afterHookContext = HookContext.from(
+                    key, type, this.getMetadata(), provider.getMetadata(), mergeEvaluationContext(ctx), defaultValue);
```
Comment (chrfwow) [inline src/main/java/dev/openfeature/sdk/OpenFeatureClient.java]: I am not sure if we should set the `afterHookContext` here. The value assigned here is not the context for the after hooks and technically, it might change the context available to the error hooks if there is an error when merging the context

File: src/main/java/dev/openfeature/sdk/HookSupport.java [hunk 4]
```diff
@@ -71,26 +101,51 @@ private <T> void executeChecked(Hook<T> hook, Consumer<Hook<T>> hookCode, String
...
 
     private EvaluationContext callBeforeHooks(
-            FlagValueType flagValueType, HookContext hookCtx, List<Hook> hooks, Map<String, Object> hints) {
+            FlagValueType flagValueType,
+            HookContext hookCtx,
+            List<Pair<Hook, HookData>> hookDataPairs,
+            Map<String, Object> hints) {
         // These traverse backwards from normal.
-        List<Hook> reversedHooks = new ArrayList<>(hooks);
+        List<Pair<Hook, HookData>> reversedHooks = new ArrayList<>(hookDataPairs);
```
Comment (chrfwow) [inline src/main/java/dev/openfeature/sdk/HookSupport.java:124]:
```suggestion
        List<Pair<Hook, HookContext>> reversedHooks = new ArrayList<>(hookDataPairs);
```
Is this possible? That way we wouldn't need to create the `HookContext` from the `HookData` every time we invoke hooks

File: src/main/java/dev/openfeature/sdk/HookSupport.java [hunk 5]
```diff
@@ -5,62 +5,92 @@
...
+            List<Pair<Hook, HookData>> hookDataPairs,
             Map<String, Object> hints) {
-        executeHooks(flagValueType, hooks, "error", hook -> hook.error(hookCtx, e, hints));
+        executeHooks(flagValueType, hookDataPairs, hookCtx, "error", (hook, ctx) -> hook.error(ctx, e, hints));
+    }
+
+    public List<Pair<Hook, HookData>> getHookDataPairs(List<Hook> hooks) {
+        var pairs = new ArrayList<Pair<Hook, HookData>>();
+        for (Hook hook : hooks) {
+            pairs.add(Pair.of(hook, HookData.create()));
```
Comment (chrfwow) [inline src/main/java/dev/openfeature/sdk/HookSupport.java]: We might be able to wrap creating the hook pair with an if
```
if (hook.supportsFlagValueType(flagValueType)) {
// create
}
```
This way we only add hooks to the list that will be run, which should also make this check in the hook execution redundant.

File: src/test/java/dev/openfeature/sdk/HookDataTest.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+        hookData.set("key", "original");
+        assertEquals("original", hookData.get("key"));
+
+        hookData.set("key", "updated");
+        assertEquals("updated", hookData.get("key"));
+    }
+
+    @Test
+    void shouldBeThreadSafe() throws InterruptedException {
```
Comment (chrfwow) [inline src/test/java/dev/openfeature/sdk/HookDataTest.java]: Please remove this test

File: src/main/java/dev/openfeature/sdk/HookSupport.java (hunk 1 above)
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/HookSupport.java]: i will use a simpe list of map.Entries or create an own class for this - pair is really easy

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java (hunk 3 above)
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/OpenFeatureClient.java]: it is a flow issue, what if we have an exception before we reach the afterhookcontext, eg. in one of the before hooks - then we do not have any kind of hook context for the error hook. I know this is not ideal, most likely due to naming. what about simple renaming to just hookContext?

File: src/main/java/dev/openfeature/sdk/HookSupport.java (hunk 4 above)
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/HookSupport.java:124]: so theory crafting, but imho worth an own follow up pr.

1. We rename the hookContext to HookContextWithoutData (strange naming, maybe we find a better one)
2. We create an interface called HookContext to enure backwards compatibility with all kind of constructors etc. and static methods (like `.from`  - not sure if we can do that)
3. We add methods for getting HookData to HookContext
4. HookContextWithoutData should implement this interface, but we do not return anything for the getters for hookdata, or an Global empty immutable object (we need to discuss this)
5. We generate a HookContextWithDataDecorator. which:
    - implements hookContext
    - has two fields: hookContext and hookData
    - delegates all methods to hookContext, except hookData
6. this way we have backwards compatibility and we do not need to duplicate the hook context multiple times.

wdyt?

File: src/main/java/dev/openfeature/sdk/HookSupport.java (hunk 5 above)
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/HookSupport.java]: really good idea! added

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java (hunk 3 above)
Comment (chrfwow) [reply src/main/java/dev/openfeature/sdk/OpenFeatureClient.java]: we can do that, but i think currently we have this behaviour were the context would be `null` in some error cases

File: src/main/java/dev/openfeature/sdk/HookSupport.java (hunk 4 above)
Comment (chrfwow) [reply src/main/java/dev/openfeature/sdk/HookSupport.java:124]: No static methods in interfaces in java, but couldn't we just make `HookContextWithDataDecorator` extend `HookContext`?

File: src/main/java/dev/openfeature/sdk/HookSupport.java (hunk 1 above)
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/HookSupport.java]: create a package private pair class

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java (hunk 3 above)
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/OpenFeatureClient.java]: should it be null in some errorcases? that is the question, because i do think the context is important (provider and state) - so i think it is valuable information even on error cases. i think the old implementation is more like a bug, and this behaviour feels "more" right

File: src/main/java/dev/openfeature/sdk/HookSupport.java (hunk 4 above)
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/HookSupport.java:124]: i am not a big fan of inheritance, and prefer composition, and i do think that this will be nicer with an interface, compared to the class, it adds a dependency where we rely on implementation of hookContext, rather than a simple interface contract

File: src/main/java/dev/openfeature/sdk/Pair.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+class Pair<K, V> {
+    private final K key;
+    private final V value;
+
+    private Pair(K key, V value) {
+        this.key = key;
+        this.value = value;
+    }
+
+    public K getKey() {
```
Comment (guidobrei) [inline src/main/java/dev/openfeature/sdk/Pair.java]: Can we just rely on `Left`/`Key` and `Right`/`Value`, and not 2 names for the same concept?

File: src/main/java/dev/openfeature/sdk/ImmutableContext.java [hunk 8]
```diff
@@ -20,6 +20,8 @@
 @SuppressWarnings("PMD.BeanMembersShouldSerialize")
 public final class ImmutableContext implements EvaluationContext {
 
+    public static final ImmutableContext EMPTY = new ImmutableContext();
```
Comment (chrfwow) [inline src/main/java/dev/openfeature/sdk/ImmutableContext.java:23]: Do we need this?

File: src/test/java/dev/openfeature/sdk/benchmark/AllocationBenchmark.java [hunk 9]
```diff
@@ -54,6 +54,30 @@ public Optional<EvaluationContext> before(HookContext<Object> ctx, Map<String, O
                 return Optional.ofNullable(new ImmutableContext());
             }
         });
+        client.addHooks(new Hook<String>() {
```
Comment (chrfwow) [inline src/test/java/dev/openfeature/sdk/benchmark/AllocationBenchmark.java:57]: Doesn't this have to be a `StringHook` for the filter to work?

File: src/main/java/dev/openfeature/sdk/HookContextWithoutData.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+     * @return resulting context for hook
+     */
+    static <T> HookContextWithoutData<T> from(
+            String key,
+            FlagValueType type,
+            ClientMetadata clientMetadata,
+            Metadata providerMetadata,
+            EvaluationContext ctx,
+            T defaultValue) {
+        return HookContextWithoutData.<T>builder()
```
Comment (guidobrei) [inline src/main/java/dev/openfeature/sdk/HookContextWithoutData.java]: You could get rid of the builder.

File: src/main/java/dev/openfeature/sdk/ImmutableContext.java (hunk 8 above)
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/ImmutableContext.java:23]: i am using it for creation of a hookcontext without context -> so i only have one instance, and not generate multiple. i thought here is the best place for it

File: src/main/java/dev/openfeature/sdk/HookData.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            data.put(key, value);
+        }
+
+        @Override
+        public Object get(String key) {
+            if (data == null) {
+                return null;
+            }
+            return data.get(key);
+        }
```
[the remark this answers is not in the document]
Comment (chrfwow) [reply src/main/java/dev/openfeature/sdk/HookData.java:67]: Will only ever be used in a single threaded context, and lazy initialization will save memory churn in case hooks do not use HookData, which will likely be the more common case
